File: android-native-shell-scaffold/app/src/main/java/com/vinesight/shell/TabsShellViewModel.kt
package com.vinesight.shell

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class TabsShellViewModel {
    private val _currentRoute = MutableStateFlow(NativeRouteId.TABS_HOME)
    val currentRoute: StateFlow<NativeRouteId> = _currentRoute.asStateFlow()

    fun select(tab: NativeTabSelection) {
        _currentRoute.value = TabsFlowCoordinator.nextRoute(
            _currentRoute.value,
            TabsFlowEvent.Select(tab),
        )
    }

    fun resetHome() {
        _currentRoute.value = TabsFlowCoordinator.nextRoute(
            _currentRoute.value,
            TabsFlowEvent.ResetHome,
        )
    }
}

File: ios-native-shell-scaffold/VineSightShell/TabsShellViewModel.swift
import Foundation
import SwiftUI

@MainActor
final class TabsShellViewModel: ObservableObject {
    @Published private(set) var currentRoute: NativeRouteId = .tabsHome

    var selectedTab: NativeTabSelection {
        TabsFlowCoordinator.resolveTab(from: currentRoute)
    }

    func select(tab: NativeTabSelection) {
        currentRoute = TabsFlowCoordinator.nextRoute(current: currentRoute, event: .select(tab: tab))
    }

    func resetHome() {
        currentRoute = TabsFlowCoordinator.nextRoute(current: currentRoute, event: .resetHome)
    }
}
